Make Nikon ContrastDetectAFInFocus tags Bytes that print as "Yes" or "No"

mallman opened the pull request:
This is the follow on PR from #2102, which was closed without merge.

This PR adds the `ContrastDetectAFInFocus` field to the `nikonAf22Def` in tiffimage_int.cpp and defines its type as a single UnsignedByte. It also changes the data type of the same field in `nikonAf21Def`.

This PR also updates the data types and print methods of the corresponding MakerNote TagInfo structures in nikonmn_int.cpp.

The `ContrastDetectAFInFocus` tag is documented on exiftool's web site: https://exiftool.org/TagNames/Nikon.html#AFInfo2.

TO DO:

- [x] Update documentation that gets uploaded to https://exiv2.org/tags-nikon.html

mallman committed:
Make Exif.NikonAf2.ContrastDetectAFInFocus and
Exif.NikonAf22.ContrastDetectAFInFocus byte types

mallman committed:
Add autogenerated documentation for NikonAf22 tags to html docs

mallman commented:
@postscript-dev @clanmills I believe I've updated the document generation to include the NikonAf22 tags in the website documentation. I tried generating the documentation locally but was not completely successful. Can someone double check the proper procedure? I'm running cmake against the "doc" target with "-DEXIV2_BUILD_DOC=ON" and it appears to complete successfully. However there is no tags-nikon.html.

On the other hand, if I build all, define the "EXIV2HOME" env var and run `make` in "doc/templates" I do get the tags-nikon.html file and it appears to be correct. I just need help verifying that this file is properly included in the documentation that's uploaded to the website.

Cheers.

mallman commented:
Of course, aside from updating the doc files in repo, how do we actually update the website's documentation?

clanmills reviewed (commented):
This looks pretty good, @mallman.  I'm going to leave it to the Team to dig more carefully through your PR and to approve your PR.  I'd like to make the following comments:

1. You've made the same change that I found in to add the the '82' to nikonAf22Cfg.
2. You've changed changed ContrastDetectAFInFocus/28 from unsignedShort from to 'unsignedByte'.  Why?  Do you know if we have a test image that uses this?
3. Your changes for `make docs` seem right to me.

exiv2.org is in transition from being hosted on AWS to GitHub and Nehal is taking over the role of "Release Engineer".  The current release procedure is documented in team.exiv2.org/website/Checklist.txt.  There will be an announcement when RC1 of the next release is available and that's the time to check that the public web-pages have been successfully modified.

4.  Thank You for fixing the test suite.  I hope you discovered the quick way to fix the regression tests which is to `$ cp test/tmp/* test/data/test_reference_files` and then clean up the git _Untracked files_. 

When working on "real" changes such as test_issue_646.py, you have to use the dentist's method _(you pull them out, one at a time)_.

mallman commented:
Hi @clanmills.

> You've changed changed ContrastDetectAFInFocus/28 from unsignedShort from to 'unsignedByte'. Why? Do you know if we have a test image that uses this?

I changed this type in both tiffimage_int.cpp and nikonmn_int.cpp because

1. From my inspection of 149 of my own files which use the nikonAf2 tag group, the bytes at offsets 28 and 29 consist of the following two byte sequences only: 0 0 and 1 0.
2. From my inspection of 16,549 of my own files which use the analogous tag group nikonAf22, the bytes at offsets 82 and 83 consist of the following two byte sequences only: 0 0 and 1 0.
3. The exiftool documentation at https://exiftool.org/TagNames/Nikon.html#AFInfo2 asserts that this is an unsigned byte.

I don't know why Nikon would define a 30 and 84 long byte array to serialize information if the last byte is always 0. Either the last entry is an unsigned short, the last byte is always 0 or there's an unknown tag at those last byte offsets.

There's another, practical reason for defining this as a byte instead of a short. As a @postscript-dev points out in https://github.com/Exiv2/exiv2/pull/2102#issuecomment-1041377021, changing this data type to a byte and the print function allows us to print the string value of this tag as a more readable "On" or "Off" than "1" or "0".

mallman commented:
Also, to address last your question in

> You've changed changed ContrastDetectAFInFocus/28 from unsignedShort from to 'unsignedByte'. Why? Do you know if we have a test image that uses this?

I don't know if we have an image. We have plenty of "exv" files with this tag.

clanmills commented:
@mallman Very thorough explanation.  Thanks.  And good to know this is already covered in the test suite.

You're probably aware that there are two ways to print values.  "Raw" `$ exiv2 -pv path` and "Translated" `$ exiv2 -pt path`.  Within Team Exiv2, we usually refer to "Translated" as the _**pretty printers**_.  So "On/Off" is your pretty printer.   I would expect the on/off pretty printer to handle handle any integer value (byte/short/long).

Can you check that "Raw" and "Translated" are different for both AF1 and AF2 files.

I haven't taken the time to understand the following, however I now know that you've survived an encounter with the mysterious manufacturer's metadata messy monster _and lived to tell the tail_:
> I don't know why Nikon would define a 30 and 84 long byte array to serialize information if the last byte is always 0.

clanmills commented:
BTW, from a metadata point of view, a .EXV and a .JPEG are _more or less_ the same thing and documented in my book.  We try to avoid adding big files to test/data.

postscript-dev reviewed (dismissed):
@mallman:
I am happy to approve your changes, but I have a few thoughts on the tags that may be of interesting to you.

Usually, the makernotes tags in ExifTool are more accurate than Exiv2. When I transcribe tags into Exiv2, I use both ExifTool's main website (e.g. [ExifTool's Nikon tags](https://www.exiftool.org/TagNames/Nikon.html)) and the ExifTool source code (e.g., [exiftool/lib/Image/ExifTool/Nikon.pm](https://github.com/exiftool/exiftool/blob/master/lib/Image/ExifTool/Nikon.pm)). Checking the source code is important as many of the conditions and caveats are not listed on the website. However, many of the conditions only apply when implementing translated output - which is optional in Exiv2.

When I compare Exiv2's `Exif.NikonAf2.*` and `Exif.NikonAf22.*` tags to ExifTool, I see that our two groups are only one group in ExifTool ([AFInfo2](https://www.exiftool.org/TagNames/Nikon.html#AFInfo2)).

In #646, you found that some of the tags (e.g., `Exif.NikonAf2.AFAreaXPosition`) were not producing the correct value in your test file. Looking at ExifTool's [AFInfo2](https://www.exiftool.org/TagNames/Nikon.html#AFInfo2) tags, I can see that they list more than one tag with the same name and [notes in ExifTool's source code](https://github.com/exiftool/exiftool/blob/537cba56936c21b729aeed404b117e31c3c6fed0/lib/Image/ExifTool/Nikon.pm#L3962) indicate that the tags are included/excluded by situation and/or AFInfo2 version number. Currently, Exiv2 is trying to distinguish the different groups based on the [length of the tag group](https://github.com/Exiv2/exiv2/blob/0b17cc31f71a06766a38f6c19f71e806381fadea/src/makernote_int.cpp#L1154) but it would be better to use the version number.

It would be possible to implement all the tags listed in ExifTool's [AFInfo2](https://www.exiftool.org/TagNames/Nikon.html#AFInfo2) group but we currently have some limitations:

1. Exiv2 requires unique tag names within a tag group. This means that we could include the multiple `AFAreaXPosition` tags but would need to rename the duplicates.
2. Exiv2 cannot use conditions to include/exclude individual tags from a tag group, only to include/exclude a group. This means that in an implementation, all the tags would always be listed including all the 'excluded' ones. To help compensate, the translated output can be filtered and could be set to "n/a".
3. Following on from 2., as Exiv2 cannot limit a tag to be read-only, any 'excluded' tags could potentially be set.

Instead, the solution to these problems is to use the same method employed by the Nikon flash groups. Exiv2 defines multiple `NikonFl*` groups, each of which represents a different Flash version number. In https://github.com/Exiv2/exiv2/blob/0b17cc31f71a06766a38f6c19f71e806381fadea/src/makernote_int.cpp#L1147 the version number is compared and the appropriate flash group is included. This same method could be used to provide `NikonAF2*` groups that only apply to a particular AF version.

I am not sure why ExifTool have not moved over to this method but perhaps they will, if more versions are discovered.

When looking at ExifTool's tags, I see that there is also an [AFInfo2V0400](https://www.exiftool.org/TagNames/Nikon.html#AFInfo2V0400) group but this is filtered based on the camera model number (`NIKON Z 9`). Some of the Sony groups already filter based on the model number (e.g., https://github.com/Exiv2/exiv2/blob/0b17cc31f71a06766a38f6c19f71e806381fadea/src/makernote_int.cpp#L1222).

mallman commented:
I'm having some second thoughts about the formatting of the ContrastDetectAFInFocus values. For something like ContrastDetectAF or PhaseDetectAF, "On" or "Off" represents a camera setting that a user can select. ContrastDetectAFInFocus represents a more general binary state which the camera records as part of the image metadata. To me, this is more like a "True" or "False" value.

There does not appear to be a "True"/"False" (or "true"/"false") pretty-printer in the exiv2 codebase (at least when I did a global text search). Is this an intentional omission?

@clanmills @postscript-dev In the absence of such a formatter, would it make more sense to simply print a 1 or 0 (with C-style boolean interpretation)? Is there a reason we shouldn't format this value as "True"/"False" (or "true"/"false")? Or would you prefer I stick with "On"/"Off" as this patch currently prints it?

mallman commented:
Re your request, @clanmills.

> Can you check that "Raw" and "Translated" are different for both AF1 and AF2 files.

I have verified this.

postscript-dev commented:
> I'm having some second thoughts about the formatting of the ContrastDetectAFInFocus values. For something like ContrastDetectAF or PhaseDetectAF, "On" or "Off" represents a camera setting that a user can select. ContrastDetectAFInFocus represents a more general binary state which the camera records as part of the image metadata. To me, this is more like a "True" or "False" value.
There does not appear to be a "True"/"False" (or "true"/"false") pretty-printer in the exiv2 codebase (at least when I did a global text search). Is this an intentional omission?

I don't have a Nikon camera, so I can't test these settings. The description of the tags is a little basic but seems to back your reasoning up. If you want to change this, that's OK.

> @clanmills @postscript-dev In the absence of such a formatter, would it make more sense to simply print a 1 or 0 (with C-style boolean interpretation)? Is there a reason we shouldn't format this value as "True"/"False" (or "true"/"false")? Or would you prefer I stick with "On"/"Off" as this patch currently prints it?

I am not at my computer, but if the `True`/`False` array is missing, you could create it by using the `On`/`Off` array as a template. Then simply update the reference in the tag's pretty printing. Most pretty printing is usually straight forward.

mallman commented:
> > @clanmills @postscript-dev In the absence of such a formatter, would it make more sense to simply print a 1 or 0 (with C-style boolean interpretation)? Is there a reason we shouldn't format this value as "True"/"False" (or "true"/"false")? Or would you prefer I stick with "On"/"Off" as this patch currently prints it?
> 
> I am not at my computer, but if the `True`/`False` array is missing, you could create it by using the `On`/`Off` array as a template. Then simply update the reference in the tag's pretty printing. Most pretty printing is usually straight forward.

I believe I've found the best approach to formatting this value. You may note in exiftool's source code that exiftool prints ContrastDetectAFInFocus as "Yes" or "No". I noticed that this formatting is already used for another nikon metadata tag. I'm going to push a new commit with this modification.

I'm also going to push a commit that changes the nikonAf2* selection logic from size-based to version-based.

mallman committed:
Print value of ContrastDetectAFInFocus as Yes/No. Change internal
selection logic for NikonAf2* from tag data length to group version

mallman commented:
My latest commit changes the value printer for ContrastDetectAFInFocus to print "Yes"/"No".

Also, I changed the logic for selecting the nikonAf2* tag group by using that group's version number instead of the tag length.

Our nikonAf2* tags support version strings "0100" and "0101". Interestingly, we have two exv files in the regression tests whose versions are "0200" and "0300" with tag lengths of 42 and 60, respectively. Because we were using the nikonAf22 tag parser for tag length 84 and nikonAf2 for all other tag lengths, we mistook those as nikonAf2 (i.e. version "0100") tags. This commit no longer parses AF2 tags for any version other than "0100" and "0101", and I've updated the regression tests appropriately.

Exiftool has some logic to handle version strings matching "0100", "0101", /^03/ and "0301" (though not "0300" exactly and nothing starting with "02"). Frankly it's a little hard to follow all the version matching conditions and whatnot.

The exv file we have with af2 version "0200" data is from a Nikon 1 AW1. The exv file we have with af2 version "0300" data is from a Nikon Z6. I'm not confident we were parsing that data correctly. I think that supporting those and other af2 versions will be a project (for someone) for another day.

clanmills commented:
Wow, Michael.  You are now staring the metadata monster between the eyes.

![Unknown](https://user-images.githubusercontent.com/529982/154833554-6a004db2-1872-4109-8e5f-972416573a19.jpeg)

I suspect that all 01xx records will support the same opening 28 bytes.  The longer records are a guess. 0200, 0300?  Nobody knows.  It's even possible that the engineers at Nikon can't explain it either.  Hardware Engineers thing that all Firmware people are stupid.  I have a different opinion.  How did I learn about the AF22?  I can't remember.

I have a Nikon D5300 which I purchased in 2014 because it has GPS.  There is a sentence in the manual: "If the camera has a vibration reduction (VR) switch, select **ON** to reduce the effects of vibration.".  Like most camera owners, I know very little about the camera.  I'll investigate the AF features after breakfast.

You've done a great job here.  As you've discovered, when you start digging, you can get stuck in the hole.  We have to accept that we can't solve everything.  So we'll wait for a user to open an issue and move forward from there.  Great Job:  @mallman and @postscript-dev Thank You Very Much.

clanmills commented:
BTW, IIRC, I used tagLength when selecting AF22 because the code accesses bytes 80+.  So yes, good idea to select the decoder using the 0100/0101 header _**and**_ be sure you have enough data.  You'll want to _keep the security researchers off your back_.  Those people will look at the code and maliciously construct a file with 0101 and 40 bytes of data.

postscript-dev commented:
@mallman 

Thanks for going the extra mile with your PR, the effort is appreciated. I will take a look at your PR tomorrow.

@clanmills 

> BTW, IIRC, I used tagLength when selecting AF22 because the code accesses bytes 80+. So yes, good idea to select the decoder using the 0100/0101 header and be sure you have enough data. You'll want to keep the security researchers off your back. Those people will look at the code and maliciously construct a file with 0101 and 40 bytes of data.

As the Nikon selector method is used in lots of places, the error would have big consequences. I will try and take a look at this tomorrow.

mallman commented:
> BTW, IIRC, I used tagLength when selecting AF22 because the code accesses bytes 80+. So yes, good idea to select the decoder using the 0100/0101 header _**and**_ be sure you have enough data. You'll want to _keep the security researchers off your back_. Those people will look at the code and maliciously construct a file with 0101 and 40 bytes of data.

I'm a little hazy on your security recommendation. Can you be more explicit?

clanmills commented:
When I wrote that, I was shooting from memory, Michael.  Here's an analysis on the code in branch 0.27-maintenance.  I'm pleased to say that my memory about this is correct and reasonable _(well to me anyway)_.

There's a selector:

```
tiffimage_int.cpp:        {    0x00b7, nikon3Id,         EXV_COMPLEX_BINARY_ARRAY(nikonAf2Set, nikonAf2Selector) },
```

When tag 0x00b7 occurs is the Nikon Maker note, this function tells the machine who should handle it:

```
    int nikonAf2Selector(uint16_t tag, const byte* /*pData*/, uint32_t size, TiffComponent* const /*pRoot*/)
    {
        int result = tag == 0x00b7 ? 0 : -1 ;
        if (result > -1 && size == 84 ) {
            result = 1;
        }
        return result;
    }
```
So, it's saying "if size == 84, use nikonAf2Set[1] == nikonAf22 (Def and Cfg)

```
    extern const ArraySet nikonAf2Set[] = {
        { nikonAf21Cfg, nikonAf21Def, EXV_COUNTOF(nikonAf21Def) },
        { nikonAf22Cfg, nikonAf22Def, EXV_COUNTOF(nikonAf22Def) },
    };
```

And he's
```
    extern const ArrayDef nikonAf22Def[] = {
        {  0, ttUndefined,     4 }, // Version
        {  4, ttUnsignedByte,  1 }, // ContrastDetectAF
        {  5, ttUnsignedByte,  1 }, // AFAreaMode
        {  6, ttUnsignedByte,  1 }, // PhaseDetectAF
        {  7, ttUnsignedByte,  1 }, // PrimaryAFPoint
        {  8, ttUnsignedByte,  7 }, // AFPointsUsed
        { 70, ttUnsignedShort, 1 }, // AFImageWidth
        { 72, ttUnsignedShort, 1 }, // AFImageHeight
        { 74, ttUnsignedShort, 1 }, // AFAreaXPosition
        { 76, ttUnsignedShort, 1 }, // AFAreaYPosition
        { 78, ttUnsignedShort, 1 }, // AFAreaWidth
        { 80, ttUnsignedShort, 1 }, // AFAreaHeight
    };
```

He's going to read a short at byte 80.  And that's safe if there are 84 bytes.

If you change the selector to say "if the first four bytes (version) and `0101`, use nikonAf22Def you are standing on very thin ice.  A _malicious_ file with `0101` and less that 82 bytes, will read beyond the end of a buffer.

When I added that code, I thought I don't know what Nikon are doing.  In the test/sample file, there are 84 bytes.  The code is safe and works on the test/sample file.

A couple of questions:
1.  Does that make sense to you?
2. Did you get tvisitor to build and work?

mallman committed:
When selecting which version of nikonAf2 metadata to parse, include the
size of the data declared in the tag header as part of the lookup

mallman commented:
> A couple of questions:
> 
> Does that make sense to you?

Hmmm... not sure, but I think I see your concern. Take a look at the latest commit I pushed to this PR and let me know if this addresses the issue you're referring to.

> Did you get tvisitor to build and work?

No, but I've found `exiv2 -pR` to suffice.

clanmills commented:
I believe your code is vulnerable to the attack I have explained.  Would you like to talk 1-to-1 on Zoom?  We can screen share and step this together in the debugger.

`$ exiv2 -pR` was a prototype for tvisitor and most the code in `-pR` is in tvisitor.cpp.  However tvisitor is much more robust (more thoroughly tested) and handles Iptc, Icc and Xmp as well as tiff.  

We had a Vietnamese GSoC student at NUS in Singapore working with us in 2013 and he asked the question "where is the metadata" and he started the `-pS` option (print structure) and wrote some documents for the wiki.  Alison and I attended his wedding in Ho Chi Minh City in 2017 and have visited him twice in Singapore.

clanmills commented:
I've "tweaked" _DSC8437.exv in the test suite and renamed it _DSC8437.jpg to enable it to be attached to this bug report.  I modified the 'count' field in the 0x00b7 from 84 to 6.  It's byte 17+1074+826+4=1921 which I discovered using tvisitor.  I edited the files Hex Fiend on macOS.

You will have to build with ASAN enabled: `cmake .. -DEXIV2_TEAM_USE_SANITIZERS=1`.  Two observations:

1.  Branch 0.27-maintenance doesn't throw an ASAN on the "malicious" file.
2.  The ASAN build on 'main' crashes with the command: `$ exiv2 -vVg exiv2`.  0.27-maintenance is OK.  I'll report this as different issue and I'm sure @kevinbackhouse will fix it!

I'm very happy to work with you to get this resolved.

![_DSC8437](https://user-images.githubusercontent.com/529982/155334405-3a075e1b-90c6-4806-9c61-d3a9bb048c77.jpg)

postscript-dev reviewed (dismissed):
@clanmills:
Your bug was a good catch.

@mallman:
This is good work, thanks.

clanmills commented:
@postscript-dev How can you approve this PR when it exposes a security issue?

I predicted a security issue and have proven my case.  I'm happy to be involved with getting this fixed, however I'm not offering to undertake the fix.  This code will have to be be fixed.

clanmills commented on src/makernote_int.cpp:
```diff
@@ -1151,15 +1154,6 @@ namespace Exiv2 {
         return aix == nullptr ? -1 : aix->idx_;
     }
 
-    int nikonAf2Selector(uint16_t tag, const byte* /*pData*/, uint32_t size, TiffComponent* const /*pRoot*/)
```
We cannot simply throw away this code.  The functionality can be included in nikonSelector.

```
    int nikonSelector(uint16_t tag, const byte* pData, uint32_t size, TiffComponent* const /*pRoot*/)
    {
        if (size < 4) return -1;
        const NikonArrayIdx* aix = find(nikonArrayIdx, NikonArrayIdx::Key(tag, reinterpret_cast<const char*>(pData), size));
        return aix == nullptr ? -1 : aix->idx_;
    }
```
This code is naive.  There is not check that there is sufficient data in the tag to be processed by the chosen selector.

clanmills commented:
I don't know GitHub well enough to block this PR.  @postscript-dev Please remove your approval for this PR.

postscript-dev commented:
@clanmills 
Appologies, I misunderstood how we intended to proceed with this PR. I am happy to take your advice on this

mallman commented:
> I don't know GitHub well enough to block this PR.

I'm going through your comments and re-examining the code, but I can help with this immediately.

You can block a PR by opening a review and submitting it with a "changes required" status.

mallman commented:
> I'm very happy to work with you to get this resolved.

@clanmills I appreciate that, and a Zoom call sounds great. Can we schedule an hour for tomorrow evening, at or after 18:00 GMT?

clanmills commented:
@mallman @postscript-dev I'm so happy with your response.  This PR isn't urgent and I understand you're both busy with your day jobs.  Please don't "Merge".

I think the function `nikonSelector()` is a good idea that we can _easily_ make safe.

Zoom:  I'm usually awake 06:00-24:00 UTC/GMT.  Let me know what works for you.
https://us02web.zoom.us/j/9129667335?pwd=cjFvWHNnMlpwOExYSE44dHFrdWtyZz09
Meeting ID: 912 966 7335
Passcode: 2ngkdc

mallman commented:
> @mallman @postscript-dev I'm so happy with your response. This PR isn't urgent and I understand you're both busy with your day jobs. Please don't "Merge".

Actually, working on this PR is most of my day job right now.

> I think the function `nikonSelector()` is a good idea that we can _easily_ make safe.

I'm working in that direction, too.

> Zoom: I'm usually awake 06:00-24:00 UTC/GMT. Let me know what works for you. https://us02web.zoom.us/j/9129667335?pwd=cjFvWHNnMlpwOExYSE44dHFrdWtyZz09 Meeting ID: 912 966 7335 Passcode: 2ngkdc

Let's say, 18:00 GMT?

clanmills commented:
18:00 GMT Thursday 2022-02-24 works for me.

The `nikonSelector` function needs a more robust size check.  Have decided which decoding function should to use, it must be certain there is sufficient data.  The code will be something like this:

```cpp
        //         if (size < 4) return -1;  // NO NO NO!
        NikonArrayIdx = .... 
        // check there is sufficient data to service the decoder:
        size_t need = aix != nullptr ? 4 : NikonArrayIdx[aix].start_ + NikonArrayIdx[aix].size_;
        if (size  < need) return -1;
        
        return aix == nullptr ? -1 : aix->idx_;
```

Here's a reminder of about NikonArrayIdx.  I'd like to review with you how you determined start_ and size_.  Another possibility is to put a test into the top of every tag handler.

```cpp
    //! Structure for an index into the array set of complex binary arrays.
    struct NikonArrayIdx {
        //! Key for comparisons
        struct Key {
            //! Constructor
            Key(uint16_t tag, const char* ver, uint32_t size) : tag_(tag), ver_(ver), size_(size) {}
            uint16_t    tag_;  //!< Tag number
            const char* ver_;  //!< Version string
            uint32_t    size_; //!< Size of the data (not the version string)
        };
        //! Comparison operator for a key
        bool operator==(const Key& key) const
        {
            return    key.tag_ == tag_
                   && 0 == strncmp(key.ver_, ver_, strlen(ver_))
                   && (size_ == 0 || key.size_ == size_);
        }

        uint16_t    tag_;   //!< Tag number of the binary array
        const char* ver_;   //!< Version string
        uint32_t    size_;  //!< Size of the data
        int         idx_;   //!< Index into the array set
        uint32_t    start_; //!< Start of the encrypted data
    };
```

clanmills commented:
Thanks for meeting on Zoom to discuss this, Michael. @mallman 

It's clear that issue #2114 (and #2115) are obscuring and preventing progress with your PR.   There's strong evidence from the exifprint application that your code works correctly.  I have the file test/data/_DSC8437.jpg in the test/data directory of your PR.  This is the file above.  It is _DSC8437.exv with the 0x7b tag length changed from 84 to 6.

```
614 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ git status
On branch nikon_ContrastDetectAFInFocus
Your branch is up to date with 'origin/nikon_ContrastDetectAFInFocus'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	../test/data/_DSC8437.jpg

nothing added to commit but untracked files present (use "git add" to track)
615 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ 
```

I have built you code as follows:

```bash
$ cmake .. -DEXIV2_TEAM_USE_SANITIZERS=1 -DEXIV2_BUILD_SAMPLES=1 ; make
```

The exiv2 application is very broken when built with ASAN.  The crash has nothing to do with your code:

```
611 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ env DYLD_LIBRARY_PATH=$PWD/lib bin/exiv2 -g af/i ../test/data/_DSC8437.jpg 
/Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/app/actions.cpp:497:25: runtime error: member call on address 0x0001059050a0 which does not point to an object of type 'Exiv2::Metadatum'
0x0001059050a0: note: object is of type 'Exiv2::Exifdatum'
 01 00 00 00  30 81 d9 01 01 00 00 00  10 23 80 05 01 00 00 00  60 44 60 05 01 00 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              vptr for 'Exiv2::Exifdatum'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/app/actions.cpp:497:25 in 
Abort trap: 6
612 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $
```

However your code is working correctly as can be demonstrated with the exifprint application.   First with the unmodified test file _DSC8437.exv, the metadata is correctly reported including ContrastDetectAFInFocus:

```
613 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ env DYLD_LIBRARY_PATH=$PWD/lib bin/exifprint ../test/data/_DSC8437.exv | grep -i af 
Exif.Nikon3.Focus                            0x0007 Ascii       7  AF-S  
Exif.NikonSi02xx.0x01af                      0x01af Byte      167  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
Exif.NikonLd3.AFAperture                     0x0005 Byte        1  37
Exif.NikonAf22.Version                       0x0000 Undefined   4  48 49 48 49
Exif.NikonAf22.ContrastDetectAF              0x0004 Byte        1  1
Exif.NikonAf22.AFAreaMode                    0x0005 Byte        1  1
Exif.NikonAf22.PhaseDetectAF                 0x0006 Byte        1  0
Exif.NikonAf22.PrimaryAFPoint                0x0007 Byte        1  0
Exif.NikonAf22.AFPointsUsed                  0x0008 Byte        7  0 0 0 0 0 0 0
Exif.NikonAf22.0x000f                        0x000f Byte       55  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
Exif.NikonAf22.AFImageWidth                  0x0046 Short       1  8256
Exif.NikonAf22.AFImageHeight                 0x0048 Short       1  5504
Exif.NikonAf22.AFAreaXPosition               0x004a Short       1  3950
Exif.NikonAf22.AFAreaYPosition               0x004c Short       1  2871
Exif.NikonAf22.AFAreaWidth                   0x004e Short       1  435
Exif.NikonAf22.AFAreaHeight                  0x0050 Short       1  360
Exif.NikonAf22.ContrastDetectAFInFocus       0x0052 Byte        1  1  <-- Your tag is reported
Exif.NikonAf22.0x0053                        0x0053 Byte        1  0
Exif.NikonAFT.AFFineTune                     0x0000 Byte        1  0
Exif.NikonAFT.AFFineTuneIndex                0x0001 Byte        1  255
Exif.NikonAFT.AFFineTuneAdj                  0x0002 SByte       1  0
Exif.NikonAFT.0x0003                         0x0003 Byte        1  0
614 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ 
```  

With the "doctored" file _DSC8437.jpg, the metadata is correctly reported with no ASAN errors.  The tag ContrastDetectAFInFocus cannot be reported because the tag length of 6 makes the data invisible:

```
612 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ env DYLD_LIBRARY_PATH=$PWD/lib bin/exifprint ../test/data/_DSC8437.jpg | grep -i af
Exif.Nikon3.Focus                            0x0007 Ascii       7  AF-S  
Exif.NikonSi02xx.0x01af                      0x01af Byte      167  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
Exif.NikonLd3.AFAperture                     0x0005 Byte        1  37
Exif.Nikon3.AFInfo2                          0x00b7 Undefined   6  48 49 48 49 1 1 <--- Correct there are 6 bytes
Exif.NikonAFT.AFFineTune                     0x0000 Byte        1  0
Exif.NikonAFT.AFFineTuneIndex                0x0001 Byte        1  255
Exif.NikonAFT.AFFineTuneAdj                  0x0002 SByte       1  0
Exif.NikonAFT.0x0003                         0x0003 Byte        1  0
613 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $
```

As this matter has never been urgent, I think we should wait for #2114 to be resolved, then rebase this PR, run the whole test suite and then review and merge.

If you prefer you PR to be reviewed and merged without waiting for #2114, that is fine with me.  However I would prefer to wait.

I intend to move the code for my book (tvisitor.cpp and associated CMakeLists.txt) to a new location.  Working with subdirectories is easy in SVN and challenging in git.

I thank you for your time and courtesy today and would like to repeat my appreciation of your contribution to Exiv2.  Your `nikonSelector` is ingenious and solves architectural issues when handling versioned tags in the Nikon MakerNote.

mallman commented:
> I thank you for your time and courtesy today and would like to repeat my appreciation of your contribution to Exiv2. Your nikonSelector is ingenious and solves architectural issues when handling versioned tags in the Nikon MakerNote.

@clanmills It was a pleasure meeting you, and I appreciate the feedback you're contributing to this PR.

To give credit where due, I want to clarify that I had no hand in writing `nikonSelector`. I just took advantage of what was there.

mallman committed:
Document origin of data size specifications for Nikon maker note tag 0xb7

postscript-dev commented:
@clanmills :
Thank you for reporting your zoom call discussion.

You may have already discussed this, but in relation to the size of the group, the same problem exists with the other entries in the `nikonArrayIdx` array. For example, if we were to create a 'doctored' `NikonFl7` group, the size bug would be triggered.

In addition, the various Sony selector functions have no checks for size. I haven't investigated if the size problem happens with every group or only with those mentioned above.

I don't think this problem needs to be fixed in this PR though.  

This PR has been useful, as it has helped us to discover several new bugs.

clanmills commented:
Thanks, Peter @postscript-dev.  Michael @mallman and I discussed the 'size' issue in makernote_int.cpp

```cpp
    //! Nikon binary array version lookup table
    constexpr NikonArrayIdx nikonArrayIdx[] = {
        // NikonSi
        { 0x0091, "0208",    0, 0,   4 }, // D80
        { 0x0091, "0209",    0, 1,   4 }, // D40
        { 0x0091, "0210", 5291, 2,   4 }, // D300
        { 0x0091, "0210", 5303, 3,   4 }, // D300, firmware version 1.10
        { 0x0091, "02",      0, 4,   4 }, // Other v2.* (encrypted)
        { 0x0091, "01",      0, 5,  NA }, // Other v1.* (not encrypted)
        // NikonCb
        { 0x0097, "0100",    0, 0,  NA },
        { 0x0097, "0102",    0, 1,  NA },
        { 0x0097, "0103",    0, 4,  NA },
        { 0x0097, "0205",    0, 2,   4 },
        { 0x0097, "0209",    0, 5, 284 },
        { 0x0097, "0212",    0, 5, 284 },
        { 0x0097, "0214",    0, 5, 284 },
        { 0x0097, "02",      0, 3, 284 },
        // NikonLd
        { 0x0098, "0100",    0, 0,  NA },
        { 0x0098, "0101",    0, 1,  NA },
        { 0x0098, "0201",    0, 1,   4 },
        { 0x0098, "0202",    0, 1,   4 },
        { 0x0098, "0203",    0, 1,   4 },
        { 0x0098, "0204",    0, 2,   4 },
        { 0x0098, "0800",    0, 3,   4 }, // for e.g. Z6/7
        { 0x0098, "0801",    0, 3,   4 }, // for e.g. Z6/7
        // NikonFl
        { 0x00a8, "0100",    0, 0,  NA },
        { 0x00a8, "0101",    0, 0,  NA },
        { 0x00a8, "0102",    0, 1,  NA },
        { 0x00a8, "0103",    0, 2,  NA },
        { 0x00a8, "0104",    0, 2,  NA },
        { 0x00a8, "0105",    0, 2,  NA },
        { 0x00a8, "0107",    0, 3,  NA },
        { 0x00a8, "0108",    0, 3,  NA },
        // NikonAf
        { 0x00b7, "0100",   30, 0,  NA },
        { 0x00b7, "0101",   84, 1,  NA },
    };
```

It can be calculated by reading the Def structure in src/iffImage_int.cpp, which would result in '83'.  Michael said that he obtained the '84' from the ExifTool web-site.  So we agreed that it could remain and he would add a comment about the origin/source of this value.

```cpp
    //! Nikon Auto Focus 22 binary array - definition
    constexpr ArrayDef nikonAf22Def[] = {
        {  0, ttUndefined,     4 }, // Version
        {  4, ttUnsignedByte,  1 }, // ContrastDetectAF
        {  5, ttUnsignedByte,  1 }, // AFAreaMode
        {  6, ttUnsignedByte,  1 }, // PhaseDetectAF
        {  7, ttUnsignedByte,  1 }, // PrimaryAFPoint
        {  8, ttUnsignedByte,  7 }, // AFPointsUsed
        { 70, ttUnsignedShort, 1 }, // AFImageWidth
        { 72, ttUnsignedShort, 1 }, // AFImageHeight
        { 74, ttUnsignedShort, 1 }, // AFAreaXPosition
        { 76, ttUnsignedShort, 1 }, // AFAreaYPosition
        { 78, ttUnsignedShort, 1 }, // AFAreaWidth
        { 80, ttUnsignedShort, 1 }, // AFAreaHeight
        { 82, ttUnsignedByte,  1 }, // ContrastDetectAFInFocus
    };
```

I've thought _(while asleep)_ about accepting the PR immediately without waiting for #2114 to be fixed.  I've built Michael's code with -DEXIV2_TEAM_USE_SANITIZERS=0.  It fails the regression tests and is otherwise OK.

```bash
1161 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ ctest
Test project /Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build
    Start 1: bashTests
1/6 Test #1: bashTests ........................   Passed    6.25 sec
    Start 2: bugfixTests
2/6 Test #2: bugfixTests ......................   Passed    4.51 sec
    Start 3: lensTests
3/6 Test #3: lensTests ........................   Passed    4.63 sec
    Start 4: tiffTests
4/6 Test #4: tiffTests ........................   Passed    0.08 sec
    Start 5: versionTests
5/6 Test #5: versionTests .....................   Passed    0.06 sec
    Start 6: regressionTests
6/6 Test #6: regressionTests ..................***Failed    2.20 sec

83% tests passed, 1 tests failed out of 6

Total Test time (real) =  17.75 sec

The following tests FAILED:
	  6 - regressionTests (Failed)
Errors while running CTest
Output from these tests are in: /Users/rmills/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
1162 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ 
```

My judgement is to wait for #2114 and #2117 to be serviced before merging this PR.  If you and Michael want to merge without delay, that's OK with me when the ctest/regression message has been investigated.  For sure the output of exiv2 with Michael's changes look good using the test file that's been in the code base for years (and provided by Michael).

```
1171 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ git log test/data/_DSC8437.exv 
commit b02c80f7e4fb0896ab6df6ac9af6f2521308932b
Author: Robin Mills <robin@clanmills.com>
Date:   Fri Jun 7 17:41:56 2019 +0100

    fix646_NikonAF22
1172 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ 
```

```
1164 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ env DYLD_LIBRARY_PATH=$PWD/lib bin/exiv2 -g AF/i ../test/data/_DSC8437.exv
Exif.NikonLd3.AFAperture                     Byte        1  F2.9
Exif.NikonAf22.Version                       Undefined   4  1.01
Exif.NikonAf22.ContrastDetectAF              Byte        1  On
...
Exif.NikonAf22.AFAreaHeight                  Short       1  360
Exif.NikonAf22.ContrastDetectAFInFocus       Byte        1  Yes   <--- Yup, it's here
Exif.NikonAFT.AFFineTune                     Byte        1  Off
...
1165 rmills@rmillsm1:~/gnu/github/exiv2/nikon_ContrastDetectAFInFocus/build $ 
```

postscript-dev commented:
@clanmills :
If you think it best, then I am content to wait for #2114 and #2117 to be resolved before approving this PR. As long as this is acceptable to @mallman . Either way, I agree, fixing the regression test is a priority.

Do you think it is better to fix the other size bugs (e.g., in `nikonArrayIdx`) first too?

I would like to help with these problems but sadly I don't have time. I have already comitted myself to work on other issues and need to finish them. However, if you still need me to review this PR, when ready, then let me know.

mallman commented:
> @clanmills : If you think it best, then I am content to wait for #2114 and #2117 to be resolved before approving this PR. As long as this is acceptable to @mallman . Either way, I agree, fixing the regression test is a priority.

I don't think #2114 or #2117 should block this PR. Is there something in this PR that's introduced a new bug or problem?

> Do you think it is better to fix the other size bugs (e.g., in `nikonArrayIdx`) first too?

Please no. I can see that greatly expanding the scope of this PR. Updating other values in `nikonArrayIdx` should be another issue.

postscript-dev commented:
@mallman:
> I don't think https://github.com/Exiv2/exiv2/issues/2114 or https://github.com/Exiv2/exiv2/issues/2117 should block this PR. Is there something in this PR that's introduced a new bug or problem?

There are no new problems that would block this PR. The regression test failures mentioned above can be dealt with in a different PR.

@clanmills:
> Gentlemen: This PR is beginning to look like the Ukraine - an exploding mess that could go on for years! I think we should review and merge this PR.

As we are all in agreement, I will approve and someone else can then merge. The other problems can be dealt with in other PRs.

> The arm64/vptr issue reported in https://github.com/Exiv2/exiv2/issues/2114 and https://github.com/Exiv2/exiv2/issues/2115 has nothing to do with the Nikon MakerNote AF data. I thought https://github.com/Exiv2/exiv2/issues/2114 and https://github.com/Exiv2/exiv2/issues/2115 would an easy code change. They are not.

I was referring to the other entries in `nikonArrayIdx` (e.g., `NikonFl7`) but again, I think that these can be dealt with in a different PR.

postscript-dev reviewed (approved):
@mallman:
Thank you for your contributions.

clanmills commented:
I'd also like to add my appreciation for Michael's contribution. I hope he'll join Team Exiv2.  So, while I could say "We'll meet again", I'll rephrase that to "I wish you all the best with Exiv2 as I take my bow and _finally_ retire".

I'm going to merge this.  I see it's failing the new PVS check due to username/password issue.

mallman commented:
Thanks everyone! @postscript-dev @clanmills
